### PR TITLE
refactor expectRevert

### PR DESCRIPTION
**Pull request by anvacaru**
Fixes: #1485 
Change the expectRevert rules, to support internal calls.
Previously, the `expectRevert` rules would apply only when the top of the K cell would contain the `#halt ~> #return _ _` productions. Since `#return _ _` would only be present for external calls, ignoring internal calls and contract creation calls.

### CODE REVIEW

File: include/kframework/foundry.md [hunk 1]
```diff
@@ -213,20 +213,24 @@ module FOUNDRY-CHEAT-CODES
...
  // ------------------------------------------------------------------
     rule <k> #assume(B) => . ... </k> ensures B
 
-    rule <k> #assert(false) => . ... </k>
-         <account>
-           <acctID> #address(FoundryCheat) </acctID>
-           <storage> STORAGE => STORAGE [ #loc(FoundryCheat . Failed) <- 1 ] </storage>
-           ...
-         </account>
+    rule <k> #assert(false) => #markAsFailed ... </k>
```
Comment (anvacaru) [inline include/kframework/foundry.md:224]: This rule is not used anywhere and confuses. It should be removed.

File: include/kframework/foundry.md [hunk 2]
```diff
@@ -532,46 +532,42 @@ Ignore all cheat code calls which take place while `expectRevert` is active.
...
-         </expected>
-      requires (EXPECTED =/=K .ByteArray andBool EXPECTED ==K #range(OUT, 4, #sizeByteArray(OUT) -Int 4))
-        orBool EXPECTED ==K .ByteArray
+         <expectedRevert>
+           <isRevertExpected> true </isRevertExpected>
+           <expectedDepth> ED </expectedDepth> 
+           ...
+         </expectedRevert>
+      requires CD <=Int ED
+       andBool SC =/=K EVMC_SUCCESS
```
Comment (ehildenb) [inline include/kframework/foundry.md:553]: Sholud it be `SC =/=K EVMC_SUCCESS` or `SC ==K EVMC_REVERT`? If it ends with out of gas instead, for example, should we count that as a revert?

File: include/kframework/foundry.md [hunk 3]
```diff
@@ -212,21 +212,21 @@ module FOUNDRY-CHEAT-CODES
...
-           ...
-         </account>
+     syntax KItem ::= "#markAsFailed" [klabel(foundry_markAsFailed)]
+  // ---------------------------------------------------------------
+     rule <k> #markAsFailed => . ... </k>
+          <account>
+             <acctID> #address(FoundryCheat) </acctID>
+             <storage> STORAGE => STORAGE [ #loc(FoundryCheat . Failed) <- 1 ] </storage>
+             ...
+           </account>
```
Comment (ehildenb) [inline include/kframework/foundry.md:229]: Is this how Foundry works? It marks as failed for when an expectRevert fails? Or does it have a separate way of storing that?

Also, if we are marking as failed like this, do we need the `ISREVERTEXPECTED` to be passed into `foundry_success` anymore?

File: include/kframework/foundry.md (hunk 2 above)
Comment (anvacaru) [reply include/kframework/foundry.md:553]: The only instance of the `Revert` status could that I could find in their rust implementation is here https://github.com/foundry-rs/foundry/blob/master/evm/src/executor/inspector/cheatcodes/mod.rs#L489-L505. I also looked at their `handle_expect_revert` function, but I didn't see any match for the actual status code. But my Rust knowledge is limited.

File: include/kframework/foundry.md (hunk 3 above)
Comment (anvacaru) [reply include/kframework/foundry.md:229]: Well, `expectRevert` looks at the next call made after the invocation of the cheat code. If I'm not setting the storage location, then any upcoming revert call could clear the `ISREVERTEXPECTED`, at least that was my understanding.

>Also, if we are marking as failed like this, do we need the ISREVERTEXPECTED to be passed into foundry_success anymore?

The `ISREVERTEXPECTED` is still passed to the `foundry_success` predicate in the case in which the test ends with a pending `expectRevert` (i.e. the cheat code is invoked but no further call is made).

**Pull request merged**